lang/java/avro/src/main/java/org/apache/avro/io/BinaryDecoder.java

```diff
@@ -329,9 +329,10 @@ public void skipString() throws IOException {
 
   @Override
   public ByteBuffer readBytes(ByteBuffer old) throws IOException {
-    int length = readInt();
+    long length = readLong();
```

Comment:
I would say yes to that and we should have at least one test that shows the combination of the two work.
I believe the Java versions should be able to read a file with this and also be able to write the same data to a new file with the exact same data in it (i.e. also needing the long).

Reply:
so right now both string and bytes use writeInt, because there is no situation in which they can write long for this library.

We can absolutely show writeInt can be read with readLong, but I believe that will just be a re-testing of the underlying zigzag variable length encoding. By definition Ints and Longs of the same value have equal encodings. (This could be added as a test comparing outputs of `org.apache.avro.io.BinaryData#encodeInt` and `org.apache.avro.io.BinaryData#encodeLong`)

This change will only fix the error returned to the user when trying to read an exceptionally long bytes array.

Reply:
I've added this test. Let me know what you think.

Reply:
Good test to check limits, but BinaryDecoder.readLong does not use BinaryData.encodeLong (it just share same algorithm)

Why not a more direct test like :
```java
  @Test
  public void testReadWrite() throws IOException {
    ByteArrayOutputStream out = new ByteArrayOutputStream();
    DirectBinaryEncoder encoder = new DirectBinaryEncoder(out);
    ByteBuffer buffer = ByteBuffer.wrap("Hello".getBytes(StandardCharsets.UTF_8));
    encoder.writeBytes(buffer);

    DirectBinaryDecoder decoder = new DirectBinaryDecoder(new ByteArrayInputStream(out.toByteArray()));
    ByteBuffer readBuffer = decoder.readBytes(null);
    Assert.assertArrayEquals(buffer.array(), readBuffer.array());
  }
```

Reply:
My understanding is that this is checked in anywhere that has a serde cycle with a bytes type. A good example being:
`org.apache.avro.io.TestBinaryDecoder` which reads random data with a bytes field inside a record. Which does the serde cycle used above. 

If we want to test the underlying algorithm implementation which should always allow writeInt to be read by readLong, I can modify the test I have already made although I believe it would be logically equivalent in its guarantees.